src/main/java/org/tikv/common/region/AbstractRegionStoreClient.java

```diff
@@ -110,215 +108,200 @@ public boolean onNotLeader(TiRegion newRegion) {
     blockingStub = TikvGrpc.newBlockingStub(channel);
-    asyncStub = TikvGrpc.newStub(channel);
-    if (targetStore.getProxyStore() != null) {
+    asyncStub = TikvGrpc.newFutureStub(channel);
+    if (store.getProxyStore() != null) {
       Metadata header = new Metadata();
-      header.put(TiConfiguration.FORWARD_META_DATA_KEY, targetStore.getStore().getAddress());
+      header.put(TiConfiguration.FORWARD_META_DATA_KEY, store.getStore().getAddress());
       blockingStub = MetadataUtils.attachHeaders(blockingStub, header);
       asyncStub = MetadataUtils.attachHeaders(asyncStub, header);
     }
   }
 
-  private boolean retryOtherStoreByProxyForward() {
-    if (!conf.getEnableGrpcForward()) {
-      return false;
+  private Metapb.Peer switchLeader() {
+    List<SwitchLeaderTask> responses = new LinkedList<>();
+    for (Metapb.Peer peer : region.getFollowerList()) {
+      ByteString key = region.getStartKey();
+      TiStore store = regionManager.getStoreById(peer.getStoreId());
+      ManagedChannel channel =
+          channelFactory.getChannel(
+              store.getAddress(), regionManager.getPDClient().getHostMapping());
+      TikvGrpc.TikvFutureStub stub =
+          TikvGrpc.newFutureStub(channel).withDeadlineAfter(timeout, TimeUnit.MILLISECONDS);
+      Kvrpcpb.RawGetRequest rawGetRequest =
+          Kvrpcpb.RawGetRequest.newBuilder()
+              .setContext(region.getReplicaContext(peer))
+              .setKey(key)
+              .build();
+      ListenableFuture<Kvrpcpb.RawGetResponse> task = stub.rawGet(rawGetRequest);
+      responses.add(new SwitchLeaderTask(task, peer));
     }
-    if (retryForwardTimes >= region.getFollowerList().size()) {
-      // If we try to forward request to leader by follower failed, it means that the store of old
-      // leader may be
-      // unavailable but the new leader has not been report to PD. So we can ban this store for a
-      // short time to
-      // avoid too many request try forward rather than try other peer.
-      if (originStore != null) {
-        originStore.forwardFail();
+    while (true) {
+      try {
+        Thread.sleep(2);
+      } catch (InterruptedException e) {
+        throw new GrpcException(e);
       }
-      return false;
-    }
-    TiStore proxyStore = switchProxyStore();
-    if (proxyStore == null) {
-      logger.warn(
-          String.format(
-              "no forward store can be selected for store [%s] and region[%d]",
-              targetStore.getStore().getAddress(), region.getId()));
-      if (originStore != null) {
-        originStore.forwardFail();
-      } else {
-        targetStore.forwardFail();
+      List<SwitchLeaderTask> unfinished = new LinkedList<>();
+      for (SwitchLeaderTask task : responses) {
+        if (task.task.isDone()) {
+          try {
+            Kvrpcpb.RawGetResponse resp = task.task.get();
+            if (resp != null) {
```

Comment:
no need to check whether `resp` is null since you will catch all the exceptions happened in this code block